Extracted `messageEdit` and `messageEditCancel` from `.mes_edit` and `.mes_edit_cancel`

DeclineThyself opened the pull request:
This is barely under 200 lines.
Is this okay, or would you prefer I split this in two?

Commit:
Extracted `messageEdit` and `messageEditCancel` from `.mes_edit` and `.mes_edit_cancel`

DeclineThyself commented:
/gemini review

DeclineThyself commented:
My editor's diff shows less changed lines. I'll fix it.

DeclineThyself commented on public/script.js:
```diff
@@ -7191,6 +7191,101 @@ function messageEditAuto(div) {
+
+    updateEditArrowClasses();
+}
+
+/**
+ * Close the open message editor.
+ * This deletes the user's unsaved changes.
+ * @param {number} [messageId=this_edit_mes_id]
+ */
+async function messageEditCancel(messageId = this_edit_mes_id) {
+    let text = chat[messageId]['mes'];
+
+    const thisMesDiv = $(this) ?? chatElement.children().filter(`[mesid="${messageId}"]`);
+    const thisMesBlock = thisMesDiv.closest('.mes_block') ?? thisMesDiv.find('.mes_block');
+    thisMesBlock.find('.mes_text').empty();
+    thisMesDiv.closest('.mes_edit_buttons').css('display', 'none');
+    thisMesBlock.find('.mes_buttons').css('display', '');
+    thisMesBlock.find('.mes_text')
+        .append(messageFormatting(
+            text,
+            this_edit_mes_chname,
+            chat[messageId].is_system,
+            chat[messageId].is_user,
+            messageId,
+            {},
+            false,
+        ));
+    appendMediaToMessage(chat[messageId], thisMesDiv);
+    addCopyToCodeBlocks(thisMesDiv);
+
+    const reasoningEditDone = thisMesBlock.find('.mes_reasoning_edit_cancel:visible');
+    if (reasoningEditDone.length > 0) {
+        reasoningEditDone.trigger('click');
+    }
+
+    await eventSource.emit(event_types.MESSAGE_UPDATED, messageId);
+    messageId = undefined;
+
+    showSwipeButtons();
+}
```
I tend to make the most mistakes on the simplest parts...

Cohee1207 commented:
> Is this okay, or would you prefer I split this in two?

It's fine, I can work with that. And it's a soft limit anyway.

DeclineThyself commented:
> > Is this okay, or would you prefer I split this in two?
> 
> It's fine, I can work with that. And it's a soft limit anyway.

Thank you.
Next time I'll mark this as a draft until I've reviewed with Gemini.

> My editor's diff shows less changed lines. I'll fix it.

Fixed. I had white space differences hidden.

DeclineThyself commented:
/gemini review

DeclineThyself commented:
/gemini review

Wolfsblvt commented:
@codex review

Wolfsblvt commented:
> Thank you. Next time I'll mark this as a draft until I've reviewed with Gemini.

You can switch a PR to draft and back to ready anytime yourself. Is helpful to signal you are currently still working on it.

Commit:
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4633#discussion_r241505
https://github.com/SillyTavern/SillyTavern/pull/4633#pullrequestreview-3316588180

DeclineThyself commented on public/script.js:
```diff
@@ -7191,6 +7191,110 @@ function messageEditAuto(div) {
+
+    const thisMesBlock = thisMesDiv.closest('.mes_block') ?? thisMesDiv.find('.mes_block');
+    thisMesBlock.find('.mes_text').empty();
+    thisMesDiv.closest('.mes_edit_buttons').css('display', 'none');
```
Fixed.

DeclineThyself commented:
/gemini review

Commit:
Fixed..
https://github.com/SillyTavern/SillyTavern/pull/4633#issuecomment-3383321321
https://github.com/SillyTavern/SillyTavern/pull/4633#pullrequestreview-3316602505

DeclineThyself commented on public/script.js:
```diff
@@ -7191,6 +7191,110 @@ function messageEditAuto(div) {
+
+    const thisMesBlock = thisMesDiv.closest('.mes_block') ?? thisMesDiv.find('.mes_block');
+    thisMesBlock.find('.mes_text').empty();
+    thisMesDiv.find('.mes_edit_buttons').css('display', 'none');
```
Fixed.

DeclineThyself commented:
/gemini review

DeclineThyself commented:
/gemini review

DeclineThyself commented:
...
/gemini review

DeclineThyself commented on public/script.js:
```diff
@@ -7191,6 +7191,110 @@ function messageEditAuto(div) {
+        text = text.trim();
+    }
+    thisMesText.append(
+        '<textarea id=\'curEditTextarea\' class=\'edit_textarea mdHotkeys\'></textarea>',
```
Is this out of scope?

Cohee1207 replied:
Ignore this.

Wolfsblvt replied:
Gemini stupid, because it always checks too much.

DeclineThyself commented on public/script.js:
```diff
@@ -7191,6 +7191,110 @@ function messageEditAuto(div) {
+        thisMesDiv = chatElement.children().filter(`[mesid="${messageId}"]`);
+    }
+
+    const thisMesBlock = thisMesDiv.find('.mes_block') ?? thisMesDiv.find('.mes_block');
```
Fixed.

Commit:
Fixed...

DeclineThyself commented:
/gemini review

DeclineThyself commented on public/script.js:
```diff
@@ -10225,55 +10328,9 @@ jQuery(async function () {
-            if (Number(this_edit_mes_id) === chat.length - 1) {
-                chatElement.scrollTop(chatScrollPosition);
-            }
+            var edit_mes_id = Number($(this).closest('.mes').attr('mesid'));
```
@Cohee1207 

I believe `edit_mes_id` will never be defined in this part of the code. Is this a bug?
```javascript
let chatScrollPosition = chatElement.scrollTop();
if (this_edit_mes_id !== undefined) {
    let mes_edited = chatElement.find(`[mesid="${this_edit_mes_id}"]`).find('.mes_edit_done');
    if (Number(edit_mes_id) == chat.length - 1) { //if the generating swipe (...)
        let run_edit = true;
        if (chat[edit_mes_id]['swipe_id'] !== undefined) {
            if (chat[edit_mes_id]['swipes'].length === chat[edit_mes_id]['swipe_id']) {
                run_edit = false;
            }
        }
        if (run_edit) {
            hideSwipeButtons();
        }
    }
    await messageEditDone(mes_edited);
}
```

Cohee1207 replied:
I have no idea what that part should actually do. Hiding swipes if generating?

DeclineThyself replied:
Should I leave it as is?

Wolfsblvt replied:
Lemme see if one of the LLMs can figure out what it does in context. Interesting challenge. I'll come back with results.

DeclineThyself replied:
Use this as context instead?

https://github.com/TavernAI/TavernAI/blob/main/public/script.js#L5490

```javascript
  $(document).on('click', '.mes_edit', function(){
        if(Characters.selectedID == undefined){
            return;
        }
        let run_edit = true;
        const root = messageRoot($(this));
        if(!root) { return; }
        const edit_mes_id = root ? parseInt(root.attr("mesid")) : NaN;
        if(isNaN(edit_mes_id)) { return; }
        if(this_edit_mes_id !== undefined) { return; }

        if(edit_mes_id == count_view_mes-1){ //if the generating swipe (...)
            if(chat[edit_mes_id]['swipe_id'] !== undefined){
                if(chat[edit_mes_id]['swipes'].length === chat[edit_mes_id]['swipe_id']){
                    run_edit = false;
                }
            }
            if(run_edit){
                hideSwipeButtons();
            }
        }
 ```
 
 https://github.com/SillyTavern/SillyTavern/blob/edd41989fd550a8d111fb7167d456c5614a3a610/public/script.js#L7846
 ```javascript
 $(document).on("click", ".mes_edit", async function () {
    if (this_chid !== undefined || selected_group) {
        // Previously system messages we're allowed to be edited
        /*const message = $(this).closest(".mes");

        if (message.data("isSystem")) {
            return;
        }*/

        let chatScrollPosition = $("#chat").scrollTop();
        if (this_edit_mes_id !== undefined) {
            let mes_edited = $(`#chat [mesid="${this_edit_mes_id}"]`).find(".mes_edit_done");
            if (Number(edit_mes_id) == count_view_mes - 1) { //if the generating swipe (...)
                if (chat[edit_mes_id]['swipe_id'] !== undefined) {
                    if (chat[edit_mes_id]['swipes'].length === chat[edit_mes_id]['swipe_id']) {
                        run_edit = false;
                    }
                }
                if (run_edit) {
                    hideSwipeButtons();
                }
            }
            await messageEditDone(mes_edited);
```

Wolfsblvt replied:
Have fun. I was too tired to fully comprehend this. Too long.
But I think it's actually kind of right, if I remember.

# GPT5 (medium reasoning)
I'll open the surrounding section of `public/script.js` to get full context, and search the repo for related functions and variables like `hideSwipeButtons`, `messageEditDone`, and `swipe_id` so I can explain the behavior precisely.

I'll search the codebase for related symbols ([hideSwipeButtons](cci:1://file:///d:/code/SillyTavern/public/script.js:7937:0-7941:1), [messageEditDone](cci:1://file:///d:/code/SillyTavern/public/script.js:7193:0-7228:1), `swipe_id`, `mes_edit_done`) to understand the flow and what side effects this block triggers before explaining it precisely.

I'll search for where `this_edit_mes_id` and `edit_mes_id` are defined and used to be precise about their roles, and also locate [hideSwipeButtons()](cci:1://file:///d:/code/SillyTavern/public/script.js:7937:0-7941:1) and [messageEditDone()](cci:1://file:///d:/code/SillyTavern/public/script.js:7193:0-7228:1) implementations or references if any exist elsewhere in the repo.







# What this block does

Location: [public/script.js](cci:7://file:///d:/code/SillyTavern/public/script.js:0:0-0:0) lines `10213–10227`

```js
if (this_edit_mes_id !== undefined) {
    let mes_edited = chatElement.find(`[mesid="${this_edit_mes_id}"]`).find('.mes_edit_done');
    if (Number(edit_mes_id) == chat.length - 1) { //if the generating swipe (...)
        let run_edit = true;
        if (chat[edit_mes_id]['swipe_id'] !== undefined) {
            if (chat[edit_mes_id]['swipes'].length === chat[edit_mes_id]['swipe_id']) {
                run_edit = false;
            }
        }
        if (run_edit) {
            hideSwipeButtons();
        }
    }
    await messageEditDone(mes_edited);
}
```

## Conceptual explanation

- **Single active editor**  
  This is the “clean up previous edit before starting a new one” guard. If you click the edit button on a message while some other message is already in edit mode (`this_edit_mes_id !== undefined`), the code first finalizes the existing edit, then proceeds to open the new one.

- **Special handling for the last chat message with swipes**  
  If the message currently being edited is the last message in the conversation (`Number(edit_mes_id) == chat.length - 1`), the code decides whether to temporarily hide the swipe UI before finalizing the edit.

  - If that last message has swipes (`swipe_id` present) and the currently selected “swipe position” is exactly at `swipes.length`, that means the user is sitting on the special “new generation slot” (i.e., swiped to the right where there is no existing swipe and the UI would show a “…” placeholder to generate a new one).
    - In this specific case, it sets `run_edit = false` → do not hide the swipe buttons. Rationale: you’re on the “generate new swipe” pseudo-swipe; hiding swipe controls while a generation slot is active would be jarring or could interfere with the ongoing/expected generation UX.

  - Otherwise (no `swipe_id`, or you’re on a real existing swipe index), it keeps `run_edit = true` and calls [hideSwipeButtons()](cci:1://file:///d:/code/SillyTavern/public/script.js:7937:0-7941:1) to temporarily remove the swipe controls before closing the edit. Rationale: avoid UI conflicts/flicker with swipe controls while the edit is being finalized.

- **Finalize the previous edit**  
  It finds the “Done” control for the currently edited message (`.mes_edit_done`) and calls `await messageEditDone(mes_edited)` to commit/save the edit and exit edit mode for that message.

## How it fits in the flow around it

Immediately after this block, the handler switches the clicked message into edit mode:

- Clears the message text area and toggles `.mes_buttons`/`.mes_edit_buttons` (`10228–10231`).
- Sets up a new `edit_mes_id` and updates `this_edit_mes_id` to the newly clicked message (`10231–10233`).
- Optionally triggers reasoning edit if present (`10234–10238`).
- Loads and optionally trims the message text, injects the `<textarea>`, focuses it, sets cursor at the end, and restores scroll if needed (`10240–10274`).
- Calls [updateEditArrowClasses()](cci:1://file:///d:/code/SillyTavern/public/script.js:8076:0-8094:1) to refresh any edit-related UI affordances (`10276–10277`).

## In plain terms

- **You click “Edit” on a message while another message is already being edited.**  
- The app first closes and saves the currently edited message.  
- If that previously edited message is the very last message:
  - And you’re on a real, existing swipe → hide swipe arrows before closing edit (cleaner UI).
  - And you’re on the special “generate a new swipe” slot (the “…” to the right) → keep swipe arrows visible (don’t disrupt ongoing/expected generation).
- Then it proceeds to open the editor for the message you just clicked, without messing up the scroll.

## Why this matters (UX intent)

- **Prevents multiple concurrent editors**, which could corrupt state or confuse users.
- **Respects generation state on the last message**:
  - If you’re about to generate a new swipe (the “retry” slot), the UI stays ready for that and isn’t hidden.
  - If you’re on a normal swipe, the UI is tidied up before closing the edit.
- **Avoids scroll jumps** by capturing and restoring the scroll position when editing the last message.

In short: this block gracefully hands off edit focus from one message to another, with special care not to disrupt the “retry/new swipe” generation UI on the last message.

DeclineThyself replied:
GPT-5 did not realize that `edit_mes_id` will always be `undefined` on this line.
`if (Number(edit_mes_id) == chat.length - 1) { //if the generating swipe (...)`

Commit:
Fixed....
https://github.com/SillyTavern/SillyTavern/pull/4633#discussion_r2415124843
https://github.com/SillyTavern/SillyTavern/pull/4633#discussion_r2415124845

DeclineThyself commented on public/script.js:
```diff
@@ -7191,6 +7191,110 @@ function messageEditAuto(div) {
+        reasoningEdit.trigger('click');
+    }
+
+    var text = chat[edit_mes_id]['mes'];
```
Fixed.

DeclineThyself commented on public/script.js:
```diff
@@ -7191,6 +7191,110 @@ function messageEditAuto(div) {
+async function messageEditCancel(messageId = this_edit_mes_id) {
+    let text = chat[messageId]['mes'];
+    let thisMesDiv;
+    // If this is the button then select it's parent. Otherwise, select by messageId.
+    if (this?.classList?.contains('mes_edit_cancel')) {
+        thisMesDiv = $(this).closest('.mes');
+    } else
+    {
+        thisMesDiv = chatElement.children().filter(`[mesid="${messageId}"]`);
+    }
+
+    const thisMesBlock = thisMesDiv.find('.mes_block');
+    thisMesBlock.find('.mes_text').empty();
+    thisMesDiv.find('.mes_edit_buttons').css('display', 'none');
+    thisMesBlock.find('.mes_buttons').css('display', '');
+    thisMesBlock.find('.mes_text')
+        .append(messageFormatting(
+            text,
+            this_edit_mes_chname,
+            chat[messageId].is_system,
+            chat[messageId].is_user,
+            messageId,
+            {},
+            false,
+        ));
+    appendMediaToMessage(chat[messageId], thisMesDiv);
+    addCopyToCodeBlocks(thisMesDiv);
+
+    const reasoningEditDone = thisMesBlock.find('.mes_reasoning_edit_cancel:visible');
+    if (reasoningEditDone.length > 0) {
+        reasoningEditDone.trigger('click');
+    }
+
+    await eventSource.emit(event_types.MESSAGE_UPDATED, messageId);
+    if (messageId == this_edit_mes_id) {
+        this_edit_mes_id = undefined;
+    }
+
+    showSwipeButtons();
+}
```
I added a warning.

Wolfsblvt commented:
@codex review

DeclineThyself commented:
/gemini summary

Wolfsblvt commented on public/script.js:
```diff
@@ -7191,6 +7191,117 @@ function messageEditAuto(div) {
+ * Create the message edit UI.
+ * @param {number} edit_mes_id
+ */
+async function messageEdit(edit_mes_id) {
```
Question, maybe also for @Cohee1207.
If we move this and refactor this, does it make sense to also start a new js file and move this out of main script? Something like `message.js` or `message-handling.js`.

DeclineThyself replied:
I had intended to move the `swipe` function to a new file once it was approved. I held off to make the diff view easier to read.

Wolfsblvt replied:
if the whole function is already moved, I think having it in a different file doesn't make much of a difference.
Or a line-by-line diff external tools need to be used anyway.

DeclineThyself replied:
> if the whole function is already moved, I think having it in a different file doesn't make much of a difference. Or a line-by-line diff external tools need to be used anyway.

I agree, but `swipe` is already hard enough to read.

DeclineThyself commented:
@Cohee1207,  Would you like any changes made to this PR?

Cohee1207 commented:
I'll test it a bit later, was busy with other stuff.

Cohee1207 committed:
Fix type conversion errors

Cohee1207 committed:
Fix type error

Cohee1207 committed:
Fix being unable to close message edit on first message with escape

Cohee1207 reviewed (approved):
I fixed a couple of bugs introduced by a string-to-number conversion. I'll leave refactoring of messageEdit function implementation out of this PR for now.

Cohee1207 commented:
I need to hear an opinion from @RossAscends 
Do you think that hiding swipe buttons during message edit is a good idea?
Regardless of the intention, it seems to me that someone's workflow may be broken if we take that away.

DeclineThyself commented:
> I fixed a couple of bugs introduced by a string-to-number conversion. I'll leave refactoring of messageEdit function implementation out of this PR for now.

Thank you.
How did you find these bugs?  Do you use anything in addition to the ESLint VSCode extension?
What tests do you run before a commit?

Cohee1207 commented:
> How did you find these bugs?

Some by manual testing, some by reading the code.

RossAscends commented:
@Cohee1207  What workflow involves opening a message for editing, not closing it, and swiping? Regardless the edit gets closed/saved when swipe happens. If we are to be consistent with "generation requests and prompt editing are mutually exclusive tracks", then hiding swipes makes sense to me. Not all swipes result in a generation request (swiping back, etc), but they do all change the prompt to be sent for the next gen request.

Cohee1207 commented:
Another bug found (sorry):
Open edit on the last message, move it up, then move it back down. Swipe buttons appear, as far as I understand your intention, they shan't.

Cohee1207 commented:
I lean towards leaving the swipe buttons like they are now - visible on edit.

Commit:
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4633#issuecomment-3412364573

Another bug found (sorry):
Open edit on the last message, move it up, then move it back down. Swipe buttons appear, as far as I understand your intention, they shan't.

DeclineThyself commented:
> Another bug found (sorry): Open edit on the last message, move it up, then move it back down. Swipe buttons appear, as far as I understand your intention, they shan't.

Fixed, Thank you. https://github.com/SillyTavern/SillyTavern/pull/4633/commits/5195d8045e26c7483a1a839288091d66720e4eab

Cohee1207 committed:
showSwipeButtons: return early if message edit is open

Cohee1207 commented:
I think it's fine now.

DeclineThyself commented:
> I lean towards leaving the swipe buttons like they are now - visible on edit.

If a user swipes while editing a message, then message editor would be closed, deleting their unsaved message.

Cohee1207 commented:
> If a user swipes while editing a message, then message editor would be closed, deleting their unsaved message.

Yeah, I noticed. Not sure if that was intended, but we'll see. I'll let you know if someone complains about that 'feature' being gone

Cohee1207 commented:
> Regardless of the intention, it seems to me that someone's workflow may be broken if we take that away.

It's always [xkcd 1172](https://xkcd.com/1172/)
